cpp/src/arrow/ipc/reader.cc

```diff
@@ -104,6 +104,9 @@ class IpcComponentSource {
       return Status::IOError(
           "Buffers-pointer of flatbuffer-encoded RecordBatch is null.");
     }
+    if (buffer_index >= static_cast<int>(buffers->size())) {
+      return Status::IOError("buffer_index out of range.");
```

Comment:
I think this should be Status::IndexError

Reply:
(Not clang-formatted:)
```suggestion
      return Status::IndexError("buffer_index (", buffer_index, ")out of range; only ", buffers->size(), " buffers are available.");
```